[Serverless nav] Update footer + project settings cloud links

loixlab opened the pull request:
In this PR I've added support for a new `"cloudLinkId"` property on the navigation node definition. If the `kibana.yml` config has been set for the cloud link it will render in the left nav. If not it won't render.

<img width="712" alt="Screenshot 2023-07-17 at 08 34 20" src="https://github.com/elastic/kibana/assets/2854616/f1d12542-d652-4d2c-aa76-c23c8546c283">


I've also added support for a new type of navigation node: a top level "group" that renders as a "link" and not as an accordion header. As per the design, this allows us to render the `Developer tools` menu item.

<img width="246" alt="Screenshot 2023-07-17 at 09 33 49" src="https://github.com/elastic/kibana/assets/2854616/8ed991a8-25d0-41c7-8132-8a4e5f3aee96">

I've updated the "search" and "observability" navigation tree to include the cloud links and the developer tools link. For the "search" I've removed the developer tools that was in the body.

### How to test

* Add the following config in the `serverless.yml` file

```
xpack.cloud.performance_url: "/performance/"
xpack.cloud.users_and_roles_url: "/users-and-roles/"
```

* Start a serverless project (e.g. `yarn serverless oblt`)

Partially fixes https://github.com/elastic/kibana/issues/160942 (blocked by https://github.com/elastic/kibana/issues/162127)

loixlab committed:
Allow cloudLink to be passed in nav node

loixlab committed:
Add cloud config for new URLs

loixlab committed:
Update observability nav to include links

loixlab committed:
Add urls to cloud start contract

loixlab committed:
Add missing i18n

loixlab committed:
Add project settings group to serverless es

loixlab committed:
Don't force collapsed state by default

loixlab committed:
Improve comment

loixlab committed:
Update GetApi type to retrieve active state

loixlab committed:
Update navs to set active state in details views

loixlab committed:
Fix TS issue

loixlab committed:
Custom render group with no children with a link

loixlab committed:
Update default navigation to include Developer tools link

loixlab committed:
Fix issue when active node is at root level of the tree

loixlab committed:
Add dev tools link to search and observability

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

loixlab committed:
Update storybook to render cloud links

loixlab committed:
Add jest test for cloud links

loixlab committed:
Merge branch 'project-settings-cloud-links' of github.com:sebelga/kibana into project-settings-cloud-links

loixlab committed:
Remove old dev tools menu

loixlab committed:
Merge remote-tracking branch 'upstream/main' into project-settings-cloud-links

loixlab committed:
Add config to test

loixlab committed:
Add links in serverless.yml

sebelga commented on config/serverless.yml:
```diff
@@ -13,6 +13,8 @@ xpack.cloud.base_url: "https://cloud.elastic.co"
 xpack.cloud.billing_url: "/billing"
 xpack.cloud.organization_url: "/account"
+xpack.cloud.performance_url: "/performance/"
+xpack.cloud.users_and_roles_url: "/users-and-roles/"
```
@sixstringcode I've added those 2 links but they are probably wrong.
I leave them for now until we have the good ones.

@sebgl A heads up that those 2 links will also have to be injected by cloud.

sebgl replied:
@sebelga ack. Would it make sense for someone in your team to add these 2 lines by raising a PR in https://github.com/elastic/project-controller/blob/main/internal/application/kibana/cloud_config.go#L56 whenever the corresponding settings have been made available in Kibana and deployed in QA+staging?

pgayvallet replied:
Are those urls *every* going to change? 

`baseUrl` for sure depends on the environment, so we need that injected by the controller, but can't those static url suffixes (that are all appended to the baseUrl) just live in the our serverless config file for now?

sebelga replied:
> Would it make sense for someone in your team to add these 2 lines by raising a PR

@sebgl I can surely do that unless we decide to keep them in Kibana from Pierre's comment.

> Are those paths ever going to change?

It's a good question. Who is responsible to update those if/when cloud decides/needs to make a change to their routes? It seems to me that it should be on Cloud side. At the same time they need to be aware and remember the injection to not only update their route but also the config.

sebelga replied:
We probably want en e2e test here and make sure none of our cloud link points to a 404 in case the route would ever change. 
cc @Dosant

Dosant replied:
> We probably want en e2e test here and make sure none of our cloud link points to a 404 in case the route would ever change.

I'd like to check with @elastic/kibana-qa before considering adding something like this into an existing test_serverless suite. It must be not as straightforward as this.

sebelga replied:
> can't those static url suffixes (that are all appended to the baseUrl) just live in Kibana's serverless config file for now

@sebgl I see that you thumbed up (is that a verb? 😄 ). Does that mean that you think that Kibana should own those links and we remove them from the config file you pointed above?

I just wonder how the maintenance of those links will happen over time.

sebgl replied:
> @sebgl I see that you thumbed up (is that a verb? 😄 ). Does that mean that you think that Kibana should own those links and we remove them from the config file you pointed above? I just wonder how the maintenance of those links will happen over time.

Yeah I was thinking that having that piece of configuration live in Kibana config (removing the indirection) may actually make things simpler. In any case we cannot introduce a breaking change of those URLs just like that as the impact would be quite important. If we introduce a breaking change we'd need to coordinate it with the Kibana team in both cases I suppose.
OTOH having them in the project-controller codebase makes the breaking change more obvious/discoverable on control plane codebase side. I have mixed feelings overall 😄 I don't feel very strongly about it. It looks like at least half these URLs already live in the settings and are populated from control-plane code already, so it might be simpler to stick to the current pattern for consistency. Doing half the settings one way and half the other way is probably the worst option we have.

sebelga replied:
Yeah makes sense! 👍 
I removed the links from the Kibana `serverless.yml` (https://github.com/elastic/kibana/pull/161971/commits/de41cd2ec3fc288ac5c008c47fa9ecade747aaa2) and created an issue (https://github.com/elastic/kibana/issues/162127) to inject the links once we have the URLs

elasticmachine commented:
Pinging @elastic/appex-sharedux (Team:SharedUX)

azasypkin commented on config/serverless.yml:
```diff
@@ -13,6 +13,8 @@ xpack.cloud.base_url: "https://cloud.elastic.co"
 xpack.cloud.profile_url: "/user/settings"
 xpack.cloud.billing_url: "/billing"
 xpack.cloud.organization_url: "/account"
+xpack.cloud.performance_url: "/performance/"
+xpack.cloud.users_and_roles_url: "/users-and-roles/"
```
__question:__ any reason you use trailing slash for `performance_url` and `users_and_roles_url` and **don't** use it for `profile_url`, `billing_url`, and `organization_url` (e.g. you dynamically append more segments to some URLs, but not to the others)? How does the Cloud proxy handle trailing slashes ([redirect or rewrite or something else](https://github.com/slorber/trailing-slash-guide))?

If there is no particular reason, let's try to be consistent and pick a single approach that works the best with the Cloud proxy.

azasypkin reviewed (approved):
Changes in `core_plugins/rendering.ts` LGTM, left a question regarding trailing slashes.

sebelga commented on x-pack/plugins/serverless_search/public/layout/nav.tsx:
```diff
@@ -52,9 +46,19 @@ const navigationTree: NavigationTreeDefinition = {
             },
             {
               link: 'dashboards',
+              getIsActive: ({ pathNameSerialized, prepend }) => {
```
I added this to maintain the navigation active state when going from the list -> detailed view

sebelga commented on x-pack/plugins/serverless_search/public/layout/nav.tsx:
```diff
@@ -52,9 +46,19 @@ const navigationTree: NavigationTreeDefinition = {
             },
             {
               link: 'visualize',
+              getIsActive: ({ pathNameSerialized, prepend }) => {
```
I added this to maintain the navigation active state when going from the list to the different visualizations detailed view

Dosant reviewed (approved):
Lgtm 👍 
And nice other smaller fixes 👏

loixlab committed:
Merge remote-tracking branch 'upstream/main' into project-settings-cloud-links

loixlab committed:
Remove trailing slash

TattdCodeMonkey commented on x-pack/plugins/serverless_search/public/layout/nav.tsx:
```diff
@@ -34,13 +35,6 @@ const navigationTree: NavigationTreeDefinition = {
           }),
           link: 'serverlessElasticsearch',
         },
-        {
```
I'd want to make sure PM & UX is aware of this and ok with moving these links out of the sidebar

cc: @alvinr @julianrosado @sphilipse

TattdCodeMonkey replied:
@sebelga this change does not align with the current UX design we have for ES3, can you remove this change and the dev tools navGroup below?

sebelga replied:
Ok I will remove.

cc @sixstringcode I think we can stop talking about "reference implementation" 😊 All the team have different designs.

loixlab committed:
Merge remote-tracking branch 'upstream/main' into project-settings-cloud-links

loixlab committed:
Revert change to serverless search nav

loixlab committed:
Revert change to obs service type

kpatticha commented on packages/shared-ux/chrome/navigation/src/ui/components/group_as_link.tsx:
```diff
@@ -0,0 +1,61 @@
+
+      <EuiFlexItem>
+        {/* eslint-disable-next-line @elastic/eui/href-or-on-click */}
+        <EuiLink
```
nit: I expect the external links to open in a new tab.

sebelga replied:
The could link could open in a new tab, let's revisit later.
For now I follow the pattern of the user avatar dropdown (https://github.com/elastic/kibana/issues/159419) where the links ("billing"...) open in the same window.

kpatticha commented on x-pack/plugins/serverless_observability/public/components/side_navigation/index.tsx:
```diff
@@ -127,10 +139,20 @@ const navigationTree: NavigationTreeDefinition = {
       type: 'navGroup',
-      id: 'projest_settings_project_nav',
-      title: 'Project settings',
+      id: 'devTools',
```
I want to check with the UX team if the dev tool was intentionally left out of the observability offering mockups or if it was an oversight.

 I don't want to hold up the work, I can revisit it later if needed.

sebelga replied:
Ok cool thanks. Yeah you can always remove it if you don't need it 👍

kpatticha reviewed (approved):
oblt changes LGTM. Thanks for the additional fixes and improvements 🥳

loixlab committed:
Remove static cloud link

We'll inject them from cloud. Issue https://github.com/elastic/kibana/issues/162127

sebelga commented:
Thanks for the review @kpatticha !
